### ISSUE TITLE
Out of memory size!

### ISSUE THREAD

**Issue by skptl** [realm/realm-java#2631]
I am seeing a very strange crash for out-of memory on Samsung SM N910F. One other device seeing this issue is Intex AQUA Star II 16GB. I understand there might be a dozen of reasons that could cause out-of-memory error. But this has been reported on just this two devices.

```
Fatal Exception: io.realm.exceptions.RealmError: Unrecoverable error. mmap() failed: Out of memory size: 6291456 in io_realm_internal_SharedGroup.cpp line 188
       at io.realm.internal.SharedGroup.nativeCommitAndContinueAsRead(SharedGroup.java)
       at io.realm.internal.SharedGroup.commitAndContinueAsRead(SharedGroup.java:95)
       at io.realm.internal.ImplicitTransaction.commitAndContinueAsRead(ImplicitTransaction.java:62)
       at io.realm.internal.SharedGroupManager.commitAndContinueAsRead(SharedGroupManager.java:118)
       at io.realm.BaseRealm.commitTransaction(BaseRealm.java:319)
       at io.realm.BaseRealm.commitTransaction(BaseRealm.java:306)
       at io.realm.Realm.commitTransaction(Realm.java:120)
```

**Comment by stk1m1** [comment]
@skptl 

Could you please provide us a simple example that has caused the issue you are going through?

**Comment by skptl** [comment]
Hi @stk1m1 
Thanks for the quick response. I am seeing this issue in our production app's Crashlytics. I am not sure how will I reproduce an example for it. But it is a simple copyOrUpdate to realm and it works fine on all other devices, all crashes seen are for these two devices with 95% crashes on Samsung SM N910F.

**Comment by beeender** [comment]
@skptl How big of your Realm db file could be? Also, did you start a thread with a Realm instance in but the Realm is not closed when terminate the thread?

**Comment by skptl** [comment]
@beeender the app doesn't store too much, I would say a maximum of 10MB depending on user data. I have never seen it grow more than 4MB. It's likely that the realm doesn't close while the user is still on the same activity, but I close realm as soon as one leaves activity.

**Comment by beeender** [comment]
@skptl That is really strange. How many threads could the app have using Realm at the same time?

Realm is using mmap in the underlying storage engine, out-of-memory could only happen when the device running out of the virtual memory space. But for you case it doesn't seem to be like that since you db file is quite small.

**Comment by skptl** [comment]
@beeender That is exactly why opened up the issue as it was specific to the two devices and db size wasn't huge. There could be a handful of threads that could be writing as it receives data from network, but each thread closes realm open writing.

**Comment by beeender** [comment]
@skptl Anymore information you can see from the Crashlytics? Do you use any other native libraries which may use the virtual memory space as well?

**Comment by Zhuinden** [comment]
(I'm not from Realm) 
Don't forget that if you open a Realm on multiple threads at the same time, then each of them get a copy of the Realm as a "snapshot" at that given time due to the MVCC architecture, so if you have too many threads at the same time, you might run into issues. Try making execution serial where possible.

Quick question, did you have this issue with 0.88.3, or only with 0.89.0?

**Comment by skptl** [comment]
@beeender I will get you all the information I have in Crashlytics. Other libraries that might be using virtual memory would be render-script, fresco-library and possibly crashlytics itself.

@Zhuinden Thanks, I try to make the execution serial when possible but anytime any service results are returned it updates realm and all those calls are not serial. I haven't published the app with realm updated to 0.89.0 as there a few API changes in realm.

**Comment by Zhuinden** [comment]
@skptl  the greatest change is how iteration works while you're in a transaction on a RealmResults.

If you mess up, you can easily run out of memory by getting stuck in an infinite loop calling `.get()` on your realm results.
